### PR TITLE
Stop logging Pods without Nodes yet as an Error

### PR DESCRIPTION
**Pull request by markmandel**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup

> /kind documentation
> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:

A common debugging red herring I see is people seeing an error that reads something akin to:

```
"message":"error retrieving node  for Pod game-servercbtld: node \"\" not found","severity":"error"
```

And (not unsurprisingly, it's an `error`) wondering if it causing whatever issue they are having with Agones.

Except these errors are totally expected, and normal, as Pods often aren't populated with the Node they live on straight away, so I wanted to have a way to be able to return an error from a workerqueue Handler function to retry, but not have it be logged at an Error level.

So that's what this does - provide a wrapper Error type that only gets logged at the Debug level, and the corresponding changes to make that error message above be logged at the Debug level and not cause so much noise.

**Which issue(s) this PR fixes**:

N/A

**Special notes for your reviewer**:

If anyone has a better name than "DebugError" please suggest. I wasn't 100% happy, but couldn't come up with anything better.

### CODE REVIEW

File: pkg/util/workerqueue/workerqueue.go [hunk 1]
```diff
@@ -37,6 +37,29 @@ const (
 	workFx = time.Second
 )
 
+// DebugError is a marker type for errors that that should only be logged at a Debug level.
+// Useful if you want a Handler to be retried, but not logged at an Error level.
+type DebugError struct {
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue.go:42]: Is this type intended to be used outside of this package? Given that the constructor below returns `error` it seems like this should be package scoped and callers should use `NewDebugError` to create them.

File: pkg/util/workerqueue/workerqueue.go [hunk 2]
```diff
@@ -37,6 +37,29 @@ const (
...
+}
+
+// NewDebugError returns a DebugError wrapper around an error.
+func NewDebugError(err error) error {
+	return &DebugError{err: err}
+}
+
+// Error returns the error string
+func (l *DebugError) Error() string {
+	return l.err.Error()
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue.go:53]: Do you need to add `if l.err = nil return nil` before this line?

File: pkg/util/workerqueue/workerqueue_test.go [hunk 3]
```diff
@@ -194,3 +196,11 @@ func TestWorkerQueueEnqueueAfter(t *testing.T) {
 		assert.Fail(t, "should have got a queue'd message by now")
 	}
 }
+
+func TestIsDebugError(t *testing.T) {
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue_test.go:200]: Add a test calling `NewDebugError` with nil as an argument.

File: pkg/util/workerqueue/workerqueue.go (hunk 2 above)
Comment (markmandel) [reply pkg/util/workerqueue/workerqueue.go:53]: Doesn't hurt! Don't want the panics!

File: pkg/util/workerqueue/workerqueue_test.go (hunk 3 above)
Comment (markmandel) [reply pkg/util/workerqueue/workerqueue_test.go:200]: ah - that is a very good idea!

**Comment by markmandel** [comment]
Done! This should be good to go now. :+1:

File: pkg/util/workerqueue/workerqueue.go [hunk 4]
```diff
@@ -37,6 +37,32 @@ const (
...
+
+// NewDebugError returns a debugError wrapper around an error.
+func NewDebugError(err error) error {
+	return &debugError{err: err}
+}
+
+// Error returns the error string
+func (l *debugError) Error() string {
+	if l.err == nil {
+		return "nil"
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue.go:54]: Should this be return "<nil>"?

https://play.golang.org/p/gbFHmpOtdtJ

File: pkg/util/workerqueue/workerqueue_test.go [hunk 5]
```diff
@@ -194,3 +195,14 @@ func TestWorkerQueueEnqueueAfter(t *testing.T) {
...
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue_test.go:199]: a nit but this doesn't just test IsDebugError any longer, so it should be renamed to TestDebugError instead.

File: pkg/util/workerqueue/workerqueue_test.go [hunk 6]
```diff
@@ -194,3 +195,14 @@ func TestWorkerQueueEnqueueAfter(t *testing.T) {
...
+	err := errors.New("not a debug error")
+	assert.False(t, isDebugError(err))
+
+	err = NewDebugError(err)
+	assert.True(t, isDebugError(err))
+
+	err = NewDebugError(nil)
```
Comment (roberthbailey) [inline pkg/util/workerqueue/workerqueue_test.go:206]: What would isDebugError return for this case? I think it would still return true, since the type info would still wrap the nil error.

File: pkg/util/workerqueue/workerqueue.go (hunk 4 above)
Comment (markmandel) [reply pkg/util/workerqueue/workerqueue.go:54]: I went back and forth on that. In the rare occurrence in which it could be `nil` I thought it might be handy to be able to easily see that. Empty string so often disappears amongst other things.

Maybe it should be "<nil>" like the playground then?

File: pkg/util/workerqueue/workerqueue_test.go (hunk 6 above)
Comment (markmandel) [reply pkg/util/workerqueue/workerqueue_test.go:206]: Yep, it would, since it's a wrapper. But we should test it :+1:

**Comment by markmandel** [comment]
Hmnn. Lots of flakiness agian

```
2021-03-04 19:09:18.748 PSTStep #22 - "e2e-stable": --- FAIL: TestFleetRollingUpdate/Use_fleet_Patch_true_25% (363.90s)
Info
2021-03-04 19:09:18.748 PSTStep #22 - "e2e-stable": fleet_test.go:264: error waiting for fleet condition on fleet simple-fleet-w6szw
```

**Comment by markmandel** [comment]
```
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": --- FAIL: TestGameServerReadyAllocateReady (149.27s)
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": gameserver_test.go:511:
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": Error Trace: gameserver_test.go:511
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": Error: Received unexpected error:
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": timed out waiting for the condition
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": waiting for GameServer to be Ready 1614911819/game-serverwwmxz
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerStateWithLogger
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": /go/src/agones.dev/agones/test/e2e/framework/framework.go:260
Info
2021-03-04 18:41:49.994 PSTStep #21 - "e2e-feature-gates": agones.dev/agones/test/e2e.TestGameServerReadyAllocateReady
```

This one seems common :thinking:

**Comment by markmandel** [comment]
ERMAGHERD THAT WAS A ROUGH ONE

File: pkg/util/workerqueue/workerqueue.go (hunk 4 above)
Comment (roberthbailey) [reply pkg/util/workerqueue/workerqueue.go:54]: I meant to put "<nil>", I guess I didn't paste the middle part in though.

Comment (markmandel) [reply pkg/util/workerqueue/workerqueue.go:54]: When I go to edit your comment, I see you wrote `<nil>`, which gets escaped by Github! Cool. I'll make this change :smile:

Comment (markmandel) [reply pkg/util/workerqueue/workerqueue.go]: .....and done!

**Review by roberthbailey — APPROVED** [review]

**Comment by markmandel** [comment]
```
--- FAIL: TestGameServerAllocationDuringMultipleAllocationClients (84.88s)
    gameserverallocation_test.go:454: could not completed gsa1 allocation : an error on the server ("error updating allocated gameserver: Operation cannot be fulfilled on gameservers.agones.dev \"preferred-p5pwn-h7x4d-g2qqz\": the object has been modified; please apply your changes to the latest version and try again") has prevented the request from succeeding (post gameserverallocations.allocation.agones.dev)
    gameserverallocation_test.go:474: 
        	Error Trace:	gameserverallocation_test.go:474
        	Error:      	Not equal: 
        	            	expected: 100
        	            	actual  : 99
        	Test:       	TestGameServerAllocationDuringMultipleAllocationClients
```

That's a legit flake.
